Spigot-Server-Patches/0348-Anti-Xray.patch

```diff
@@ -0,0 +1,1581 @@
++
++        for (int i = 0; (blockData = dataPalette.getObject(i)) != null; i++) {
++            temp[i] = global[ChunkSection.GLOBAL_PALETTE.getDataBits(blockData)];
++        }
++
++        return temp;
++    }
++
++    @Override
++    public void onBlockChange(World world, BlockPosition blockPosition, IBlockData newBlockData, IBlockData oldBlockData, int flag) {
++        if (oldBlockData != null && solidGlobal[ChunkSection.GLOBAL_PALETTE.getDataBits(oldBlockData)] && !solidGlobal[ChunkSection.GLOBAL_PALETTE.getDataBits(newBlockData)] && blockPosition.getY() <= maxBlockYUpdatePosition) {
++            updateNearbyBlocks(world, blockPosition);
++        }
++    }
++
++    @Override
++    public void onPlayerLeftClickBlock(PlayerInteractManager playerInteractManager, BlockPosition blockPosition, EnumDirection enumDirection) {
++        if (blockPosition.getY() <= maxBlockYUpdatePosition) {
++            updateNearbyBlocks(playerInteractManager.world, blockPosition);
++        }
++    }
++
++    private void updateNearbyBlocks(World world, BlockPosition blockPosition) {
++        if (updateRadius >= 2) {
++            BlockPosition temp = blockPosition.west();
++            updateBlock(world, temp);
++            updateBlock(world, temp.west());
++            updateBlock(world, temp.down());
++            updateBlock(world, temp.up());
++            updateBlock(world, temp.north());
++            updateBlock(world, temp.south());
++            updateBlock(world, temp = blockPosition.east());
++            updateBlock(world, temp.east());
++            updateBlock(world, temp.down());
++            updateBlock(world, temp.up());
++            updateBlock(world, temp.north());
++            updateBlock(world, temp.south());
++            updateBlock(world, temp = blockPosition.down());
++            updateBlock(world, temp.down());
++            updateBlock(world, temp.north());
++            updateBlock(world, temp.south());
++            updateBlock(world, temp = blockPosition.up());
++            updateBlock(world, temp.up());
++            updateBlock(world, temp.north());
++            updateBlock(world, temp.south());
++            updateBlock(world, temp = blockPosition.north());
++            updateBlock(world, temp.north());
++            updateBlock(world, temp = blockPosition.south());
++            updateBlock(world, temp.south());
++        } else if (updateRadius == 1) {
++            updateBlock(world, blockPosition.west());
++            updateBlock(world, blockPosition.east());
++            updateBlock(world, blockPosition.down());
++            updateBlock(world, blockPosition.up());
++            updateBlock(world, blockPosition.north());
++            updateBlock(world, blockPosition.south());
++        } else {
++            //Do nothing if updateRadius <= 0 (test mode)
++        }
++    }
++
++    private void updateBlock(World world, BlockPosition blockPosition) {
++        if (world.isLoaded(blockPosition)) {
```

Comment:
we can do world.getTypeIfLoaded here instead and save a map lookup

Reply:
Will do.